lib/network.go

```
@@ -2351,11 +2352,101 @@ func (msg *MsgDeSoHeader) String() string {
+	VotePartialSignature *bls.Signature
+}
+
+// Byte encoder for the BlockProducerInfo with support for versioning. The encoder only
+// supports BlockProducerInfo version 1 and above. For the legacy version 0, use the
+// BlockProducerInfo.Serialize_Legacy() method instead.
+func (bpi *BlockProducerInfo) ToBytes() ([]byte, error) {
+	// BlockProducerInfo version 0 is not supported.
+	if bpi.Version == BlockProducerInfoVersion0 {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: BlockProducerInfo version 0 not supported")
+	}
+
+	encodedBytes := []byte{}
+
+	// Required Version field
+	encodedBytes = append(encodedBytes, bpi.Version)
+
+	// Required ECDSA PublicKey
+	if len(bpi.PublicKey) == 0 {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: PublicKey is required")
+	}
+	encodedBytes = append(encodedBytes, EncodeByteArray(bpi.PublicKey)...)
+
+	// The ECDSA Signature is redundant, and is removed in BlockProducerInfo version 1 and above
+
+	// Voting BLS PublicKey
+	if bpi.VotingPublicKey == nil {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: VotingPublicKey is required")
+	}
+	encodedBytes = append(encodedBytes, EncodeByteArray(bpi.VotingPublicKey.ToBytes())...)
+
+	// Vote BLS Partial Signature
+	if bpi.VotePartialSignature == nil {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: VotePartialSignature is required")
+	}
+	encodedBytes = append(encodedBytes, EncodeByteArray(bpi.VotePartialSignature.ToBytes())...)
+
+	return encodedBytes, nil
+}
+
+// Byte decoder for the BlockProducerInfo, with support for versioning. The encoder only
+// supports BlockProducerInfo version 1 and above. For the legacy version 0, use the
+// BlockProducerInfo.Serialize_Legacy() method instead.
+func (bpi *BlockProducerInfo) FromBytes(rr *bytes.Reader) error {
+	var err error
+
+	// Required Version field
+	bpi.Version, err = rr.ReadByte()
+	if err != nil {
+		return errors.Wrapf(err, "BlockProducerInfo.FromBytes: Problem reading Version")
+	}
+
```

should we throw some error is Version is 0?

---

Hmm, yeah it's fair to return an error in that case. We know the BlockProducerInfo is in an unexpected format

---

I wonder, maybe we should throw an error if `bpi.Version != BlockProducerInfoVersion1`? Potentially, if someone sets the `BlockProducerInfoVersion` to something like 2, it can conflict with future updates to `BlockProducerInfo`.

---

yeah we should validate that the version "is valid"

---

Good catch. These ToBytes/FromBytes functions should verify that `bpi.Version == BlockProducerInfoVersion1` every time and error otherwise.